####-sparse-observable.md

```diff
@@ -0,0 +1,517 @@
+
+
+## Detailed Design
+
+### Data representation of `SparseObservable`
+
+`SparseObservable` will be optimised first for memory usage.
+Its internals will be written in Rust.
+Its data representation can be thought of as being like:
+
+```rust
+// This could also be chosen to be `u64` at the cost of increased memory if
+// we decided that 4 billion bits is too restrictive.  (Note: we make the
+// `u32` trade-off for bit indices already in Qiskit.)
+type BitIndex = u32;
+
+struct SparseObservable {
+    // The number of bits the operator acts on.  This is not inferable from
+    // any other shape or values, since identities are not stored explicitly.
+    num_bits: BitIndex,
+    // The coefficients of each abstract term in in the sum.  This has as many
+    // elements as terms in the sum.
+    coeffs: Vec<Complex64>,
+    // A flat list of single-bit (s_) terms.  This is more naturally a
+    // list of lists, but is stored flat for memory usage and locality reasons.
+    s_terms: Vec<SingleBitTerm>,
+    // A flat list of the qubit indices that the corresponding entries in
+    // `s_terms` act on.
+    s_indices: Vec<BitIndex>,
+    // Indices that partition `s_terms` and `s_indices` into sublists for each
+    // individual term in the sum.  `boundaries[0]..boundaries[1]` is the range
+    // of indices into `s_terms` and `s_indices` that correspond to the first
+    // term of the sum.  All unspecified qubit indices are implicitly the
+    // identity.  This is one item longer than `coeffs`, since `boundaries[0]` is
+    // always an explicit zero (for algorithmic ease).
+    boundaries: Vec<usize>,
+}
+```
+
+The `s_terms`, `s_indices` and `boundaries` entries are somewhat similar to the CSR/CSC sparse-matrix storage format.
+
+`s_terms` _could_ most likely be bit-packed to use up half the space, since most extended alphabets being considered require 4 bits.
+The impacts of this on memory are negligible, however:
+
+* `s_indices` is the same length as `s_terms`, and the lengths of these two are the asymptotic scaling limit for memory use.
+* the scalar element of `s_indices` is a qubit index, which is likely to be chosen as `u32`, so 4 bytes.
+* the scalar element of `s_terms` is 4 bits, which requires 1 byte for alignment if not packed, or 0.5 bytes (average) if so.
+
+Consequently, the savings for bit-packing the `s_terms` vector is asymptotically _at most_ $`0.5 / 5 = 10\%`$, but in practice will be less than this.
+Bit-packing will also have some amount of performance impact, and would hinder the creation of cheap "term-view" objects within Rust space, since a term-based slice would not necessarily start on a byte-aligned `SingleBitTerm` values.
+
+The class will be able to present itself without data copying as a Python record type of a similar structure:
+
+```python
+class SparseObservableView:
+    num_bits: int
```

Comment:
Maybe a too small comment for an RFC, but should this be `num_qubits` for consistency with the other observables & circuit?

Reply:
The names and whatnot can be bikeshed, but this discrepancy was purposeful in the RFC, at least: there's interest from the primitives in not having this tied to _qubits_ because of how they (eventually) plan to work dynamic circuits into the mixture.

That said, Ian and I talked about that fairly early on in the process, and it may well be something that's better done just by primitives-side documentation, and keep `num_qubits` as the term here, as you say, for consistency.

Reply:
Calling them qubits is fine with me. The idea with dynamic circuits would be to attach observables to slices of measurement registers rather than to the entire circuit, with measurement terms prescribing instructions to insert before those registers. Therefore multiple qubits in an `SparseObservable` could in principle correspond to a single qubit, at the quantum programmer's discretion.

Since we are not there yet, and even when we get there it would be a point of minor concern, I'm not fussed.

Reply:
We can call this `qubits` in the actual implementation, no trouble.